text/0477-policy-validation.md

```diff
@@ -0,0 +1,561 @@
+to specify where the rules are located.
+
+CDK Policy Validation should never be used as a deployment gate or as an enforcement layer. It should instead be
+thought of as developer enablement. Ultimately, the development environment is under the control of individual 
+developers and development teams. It's up to them to opt in to this feature, and to ensure that the policy 
+validation is applied. The goal of CDK Policy Validation is to minimize the amount of set up needed during development, and make it as easy
+as possible. For example, a typical workflow could be:
+
+> A note on the term "deployment gate" above. It is perfectly acceptable to have
+> policy at synth run as part of you CI or CD pipeline and prevent deployments.
+> What we want to ensure is that users do not think of that "deployment gate" as
+> a way to _ensure_ that their infrastructure deployment is compliant. It should
+> instead be thought of as a way to "fail fast" and report issues before
+> deployment.
+
+* Developer writes a CDK application without the correct validations config.
+* Developer deploys non-compliant stacks.
+* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
+  application.
+* Developer adds the validations property, and avoids these violations going forward.
+
+#### Alternative
+
+```ts
+const app = new App();
+Policy.of(app).add(new CfnGuardValidator());
+```
+
+When you synthesize the CDK app the validator plugins will be called and the results will be printed.
+
+```text
+$ cdk synth
+
+Validation Report (CfnGuardValidator)
+-------------------------------------
+
+(Summary)
+
+╔═══════════╤════════════════════════╗
+║ Status    │ failure                ║
+╟───────────┼────────────────────────╢
+║ Plugin    │ CfnGuardValidator      ║
+╚═══════════╧════════════════════════╝
+
+(Violations)
+
+Ensure S3 Buckets are encrypted with a KMS CMK (2 occurrences)
+
+  Occurrences:
+  
+    - Construct Path: MyStack/MyCustomL3Construct/Bucket
+    - Stack Template Path: ./cdk.out/MyStack.template.json
+    - Creation Stack:
+        └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
+             │ Library: aws-cdk-lib/aws-s3.Bucket
+             │ Library Version: 2.50.0
+             │ Location: new Bucket (/home/hallcor/tmp/cdk-tmp-app/node_modules/aws-cdk-lib/aws-s3/lib/bucket.js:1:12889)
+             └──  MyCustomL3Construct (MyStack/MyCustomL3Construct)
+                  │ Library: N/A - (Local Construct)
+                  │ Library Version: N/A
+                  │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/custom-construct.ts:9:20)
+                  └──  MyStack (MyStack)
+                       │ Library: aws-cdk-lib.Stack
+                       │ Library Version: 2.50.0
+                       │ Location: new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)
```

Comment:
This tree is the wrong way round :) 

We can iterate on this a little. I love the idea but not the specifics of this particular rendering.

Reply:
I actually did it this direction on purpose since it shows you the violation resource first, but I'm not tied to it. The `Location` info is also a little weird, but I'm not sure what the best option is for it. For example

```
new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/custom-construct.ts:9:20)
```

This file location is actually the location of `new s3.Bucket`

Reply:
I've been convinced that the order should go top down. Updated.

Reply:
I was convinced by the argument that we should go from most familiar to the user to least familiar. At the top they see something they are familiar with and can easily trace down until they see things they are not familiar with. Bottom up they start with something they might not even know exists and then have to dig down the trace until they hit something they recognize.

If we take this example, the user probably did something like

```ts
class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
      new MyCustomL3Construct(this, 'MyCustomL3Construct');
  }
}
```

They would recognize the stack and the L3 construct that they created. They might not even know that the L3 construct creates an S3 bucket. If we show the Bucket trace first they will have to dig down the trace until the reach the L3 construct which they recognize.